File: native/ios/Property Finder/PropertyViewController.m
//
//  PropertyViewController.m
//  Property Finder
//
//  Created by Colin Eberhardt on 16/10/2012.
//  Copyright (c) 2012 Colin Eberhardt. All rights reserved.
//

#import "PropertyViewController.h"
#import "UIImage+UIImageHelpers.h"
#import "PersistentDataStore.h"
#import "AppDelegate.h"

@interface PropertyViewController ()

@end

@implementation PropertyViewController
{
    Property* _property;
    PersistentDataStore* _dataStore;
    UIImage* _starImage;
    UIImage* _noStarImage;
}

#pragma mark - initialisation

- (id)initWithProperty:(Property *)property
{
    self = [self initWithNibName:@"PropertyViewController" bundle:nil];
    if (self)
    {
        _property = property;
    }
    return self;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self)
    {
        self.title = @"Property Details";
        
        _starImage = [UIImage imageNamed:@"star.png"];
        _noStarImage = [UIImage imageNamed:@"nostar.png"];
        
        AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
        _dataStore = appDelegate.persistentDataStore;
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.priceLabel.text = _property.formattedPrice;
    self.titleLabel.text = _property.shortTitle;
    self.bedBathroomLabel.text = _property.bedBathroomText;
    self.image.image = [UIImage imageWithURLString:_property.imageUrl];
    
    self.summaryLabel.text = _property.summary;
    [self.summaryLabel sizeToFit];
    
    UIImage* buttonImage = [_dataStore isPropertyFavourited:_property] ? _starImage : _noStarImage;
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:buttonImage
                                                                              style:UIBarButtonItemStyleBordered
                                                                             target:self
                                                                             action:@selector(favouriteButtonTouched)];
    
}

- (void) favouriteButtonTouched
{
    [_dataStore toggleFavourite:_property];
    UIImage* buttonImage = [_dataStore isPropertyFavourited:_property] ? _starImage : _noStarImage;
    self.navigationItem.rightBarButtonItem.image = buttonImage;
}

@end

File: xamarin/iphone/PropertyCross/PropertyViewController.cs

using System;
using System.Drawing;

using MonoTouch.Foundation;
using MonoTouch.UIKit;

using PropertyCross.Model;
using PropertyCross.Presenter;

namespace PropertyCross
{
  public partial class PropertyViewController : UIViewController, PropertyPresenter.View
  {
    private PropertyPresenter _presenter;

    private UIBarButtonItem _favouriteButton;

    private UIImage _starImage;

    private UIImage _noStarImage;

    public PropertyViewController (PropertyPresenter presenter) : base ("PropertyViewController", null)
    {
      Title = "Property Details";
      _presenter = presenter;

      _starImage = new UIImage("star.png");
      _noStarImage = new UIImage("nostar.png");

      _favouriteButton = new UIBarButtonItem(_starImage,
            UIBarButtonItemStyle.Bordered, FavouriteButtonEventHandler);
    }
		
    public override void ViewDidLoad ()
    {
      base.ViewDidLoad ();

      NavigationItem.RightBarButtonItem = _favouriteButton;

      _presenter.SetView(this);
    }
		
    public override void ViewDidUnload ()
    {
      base.ViewDidUnload ();
      ReleaseDesignerOutlets ();
    }
		
    public override bool ShouldAutorotateToInterfaceOrientation (UIInterfaceOrientation toInterfaceOrientation)
    {
      return (toInterfaceOrientation != UIInterfaceOrientation.PortraitUpsideDown);
    }

    #region View implementation

    public event EventHandler ToggleFavourite = delegate {};

    public void SetProperty (Property property)
    {
      priceLabel.Text = property.FormattedPrice;
      titleLabel.Text = property.ShortTitle;
      descriptionLabel.Text = property.Summary;
      descriptionLabel.SizeToFit();
      bedBathroomLabel.Text = property.BedBathroomText;
      imageView.Image = UIImage.LoadFromData (NSData.FromUrl (new NSUrl (property.ImageUrl)));

    }

    public bool IsFavourited {
      set
      {
        _favouriteButton.Image = value ? _starImage : _noStarImage;
      }
    }

    #endregion

    private void FavouriteButtonEventHandler (object sender, EventArgs args)
    {
      ToggleFavourite(this, EventArgs.Empty);
    }

  }
}

